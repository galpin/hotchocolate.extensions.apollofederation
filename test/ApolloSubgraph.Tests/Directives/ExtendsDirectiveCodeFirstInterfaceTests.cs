using System.Threading.Tasks;
using HotChocolate.Types;
using Microsoft.Extensions.DependencyInjection;
using Xunit;
using static HotChocolate.Extensions.ApolloSubgraph.Test;

namespace HotChocolate.Extensions.ApolloSubgraph.Directives;

public class ExtendsDirectiveCodeFirstInterfaceTests
{
    [Fact]
    public async Task When_extends_is_specified_on_interface()
    {
        var schema = await BuildSchemaAsync(builder =>
        {
            var productInterfaceType = new InterfaceType(x =>
            {
                x.Name(nameof(IProduct));
                x.Extends();
                x.Field("upc").Key().Type<NonNullType<StringType>>();
            });
            builder.AddType(productInterfaceType);
            builder.AddObjectType<Product>(x =>
            {
                x.Implements(productInterfaceType);
                x.Extends();
                x.Field(y => y.Upc).Key().Type<NonNullType<StringType>>();
            });
            builder.AddQueryType();
        });

        var sut = schema.GetType<InterfaceType>(nameof(IProduct));

        Assert.Collection(
            sut.Directives,
            x => AssertEx.Directive(x, "extends"),
            x => AssertEx.Directive(x, "key", ("fields", "\"upc\"")));
        await schema.QuerySdlAndMatchSnapshotAsync();
    }

    [Fact]
    public async Task When_typed_and_extends_is_specified_on_interface()
    {
        var schema = await BuildSchemaAsync(builder =>
        {
            var productInterfaceType = new InterfaceType<IProduct>(x =>
            {
                x.Extends();
                x.Field("upc").Key().Type<NonNullType<StringType>>();
            });
            builder.AddObjectType<Product>(x =>
            {
                x.Implements(productInterfaceType);
                x.Extends();
                x.Field(y => y.Upc).Key().Type<NonNullType<StringType>>();
            });
            builder.AddQueryType();
        });

        var sut = schema.GetType<InterfaceType>(nameof(IProduct));

        Assert.Collection(
            sut.Directives,
            x => AssertEx.Directive(x, "extends"),
            x => AssertEx.Directive(x, "key", ("fields", "\"upc\"")));
        await schema.QuerySdlAndMatchSnapshotAsync();
    }

    private interface IProduct
    {
        string? Upc { get; }

        string? Id { get; }

        string? GetUpc();
    }

    private sealed record Product(string Upc = "1", string Id = "id") : IProduct
    {
        public string? GetUpc()
        {
            return Upc;
        }
    }
}