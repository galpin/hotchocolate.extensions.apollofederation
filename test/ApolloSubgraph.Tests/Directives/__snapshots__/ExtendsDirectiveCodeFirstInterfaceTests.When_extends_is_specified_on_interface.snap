{
  "data": {
    "_service": {
      "sdl": "interface IProduct @extends @key(fields: \u0022upc\u0022) {\n  upc: String!\n}\n\ntype Product implements IProduct @extends @key(fields: \u0022upc\u0022) {\n  upc: String!\n  id: String!\n}"
    }
  }
}
